accepted/00000-salt-ssh-with-salt-bundle.md

```diff
@@ -0,0 +1,77 @@
+
+# Detailed design
+[design]: #detailed-design
+
+## Uyuni roster
+
+We can use Uyuni roster module to provide roster data instead of creating flat roster files with Java. PR moving the logic to Uyuni roster on Java side: https://github.com/uyuni-project/uyuni/pull/4457
+
+This change can help to avoid creating temporary roster files (https://docs.saltproject.io/en/latest/topics/ssh/roster.html) and implements caching additionally.
+
+## Pre flight script
+
+Salt can execute a shell script on the salt-ssh client before the "real salt-ssh execution" starts.
+This script downloads the salt-bundle package from the appropriate bootstrap repository and extracts the virtual environment.
+By-default there is no way to pass additional parameters to pre fligt script running on the managed system.
+
+The "Pre flight script" calculates the bootstrap repository URL in a similar manner to the existing boostrap.sh script, based on the operating system and OS architecture of the system it is executed on.
+It requires one parameter from the Uyuni server: (host, port) (entry point) for SSH connections through tunnels.
+The salt codebase need to be modified to pass additional parameters, otherwise we have to generate pre flight script individually for each managed system.
+
+## Deploying Salt Bundle to the managed system
+
+We don't need to install Salt Bundle package on the managed system.
+We need to extract the virtual environment of the bundle from the package.
+Pre flight script is deploying it as described above.
+
+## Using proper salt codebase
+
+On deploying Salt Bundle with pre flight script, without modification salt-thin (`salt-thin` is a tarball with salt codebase created by the server to deploy salt on salt-ssh managed system: https://docs.saltproject.io/en/latest/ref/runners/all/salt.runners.thin.html) deployment, we will get two different codebases of salt deployed on the managed system (one codebase is from salt bundle package for the OS+arch and the other one is from salt-master delivered with salt-thin tarball).
+As the sources of these two codebases are different (update channel of the Server and client tools relevant for the system) these codebases are not in sync.
+We can prevent deploying salt-thin in case if Salt Bundle has been deployed already and use it as preferred salt codebase to handle salt-ssh session with on managed system side.
+In most cases the client tools channels are updating more frequently than the server and for the minions (not salt-ssh systems) we are using the salt codebase from client tools, it's better to use Salt Bundle codebase for salt-ssh systems.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+  * Salt Bundle consumes more space than salt-thin on managed system
+  * Salt is creating additional SSH connection to run the pre flight script 
+  * Pre flight script requires an access to the bootstrap repository (salt-thin is deployed with ssh, no http(s) connection needed for salt-ssh)
```

Comment:
What else is needed from the bootstrap repository other than the Salt Bundle package?

Reply:
`venv-enabled-ARCH.txt` to define the path to the Salt Bundle package. Pre flight script gets just these 2 files.

Reply:
Would it be possible / could it make simpler to transport those two files over the already-existing SSH channel to the minion instead of opening an http channel back to the Server?

Reply:
It requires quite significant changes in salt codebase related to salt-ssh sessions handling and impossible to do with pre flight script only. The problem is that pre flight script is used to calculate which salt bundle package should be used on base of system type and architecture. In case of using ssh do deploy these files we need few more ssh sessions to be open to the minion before, so it's more heavy from point of view of amount of open ssh connection prior the actual connection used to process the request on minion side. For this case it's better to implement something totally different than just pre flight script.

What we need:
1. Define the system type and architecture. In case of not having salt deployed yet, it's easier to do with a logic we get to the pre flight script from the bootstrap script.
2. Pass the system type and architecture to the server (salt-api in this case) and let it select the proper package to deploy on the system.
3. Extract the package on the system.

So yes, it's possible, but in this case it's better to avoid using the pre flight script feature from upstream and implement it different way with salt codebase on the server.

Reply:
I am fine with the currently described alternative, feel free to copypaste information from the response you gave above as an alternative in the "Alternatives" section.